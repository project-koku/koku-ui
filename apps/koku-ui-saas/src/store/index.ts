export { configureStore } from './store';
export type { RootState } from './rootReducer';
