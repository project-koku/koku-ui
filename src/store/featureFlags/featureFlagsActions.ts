import { createAction } from 'typesafe-actions';

export interface FeatureFlagsActionMeta {
  isCostDistributionFeatureEnabled?: boolean;
  isCostTypeFeatureEnabled?: boolean;
  isCurrencyFeatureEnabled?: boolean;
  isDefaultProjectsFeatureEnabled?: boolean;
  isExportsFeatureEnabled?: boolean;
  isIbmFeatureEnabled?: boolean;
  isNegativeFilteringFeatureEnabled?: boolean;
  isOciFeatureEnabled?: boolean;
  isUnallocatedCostsFeatureEnabled?: boolean;
}

export const setFeatureFlags = createAction('feature/init_feature_flags')<FeatureFlagsActionMeta>();
export const resetState = createAction('feature/reset_state')();
