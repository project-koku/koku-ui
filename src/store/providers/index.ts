import {
  awsProvidersQuery,
  azureProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ociProvidersQuery,
  ocpProvidersQuery,
  providersQuery,
} from 'store/providers/providersCommon';

import * as providersActions from './providersActions';
import { stateKey as providersStateKey } from './providersCommon';
import { ProvidersAction, providersReducer, ProvidersState } from './providersReducer';
import * as providersSelectors from './providersSelectors';

export {
  awsProvidersQuery,
  azureProvidersQuery,
  ociProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ocpProvidersQuery,
  ProvidersAction,
  providersActions,
  providersQuery,
  providersReducer,
  providersSelectors,
  ProvidersState,
  providersStateKey,
};
