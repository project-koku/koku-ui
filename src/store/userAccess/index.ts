import {
  awsUserAccessQuery,
  azureUserAccessQuery,
  costModelUserAccessQuery,
  gcpUserAccessQuery,
  ocpUserAccessQuery,
  userAccessQuery,
} from 'store/userAccess/userAccessCommon';

import * as userAccessActions from './userAccessActions';
import { stateKey as userAccessStateKey } from './userAccessCommon';
import type { UserAccessAction, UserAccessState } from './userAccessReducer';
import { userAccessReducer } from './userAccessReducer';
import * as userAccessSelectors from './userAccessSelectors';

export {
  userAccessQuery,
  awsUserAccessQuery,
  azureUserAccessQuery,
  costModelUserAccessQuery,
  gcpUserAccessQuery,
  ocpUserAccessQuery,
  userAccessActions,
  userAccessReducer,
  userAccessSelectors,
  userAccessStateKey,
};
export type { UserAccessAction, UserAccessState };
