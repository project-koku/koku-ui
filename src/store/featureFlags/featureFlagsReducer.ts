import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import type { resetState } from './featureFlagsActions';
import { setFeatureFlags } from './featureFlagsActions';

export type FeatureFlagsAction = ActionType<typeof setFeatureFlags | typeof resetState>;

export type FeatureFlagsState = Readonly<{
  hasFeatureFlags: boolean;
  isCostDistributionFeatureEnabled: boolean;
  isCostTypeFeatureEnabled: boolean;
  isCurrencyFeatureEnabled: boolean;
  isDefaultProjectsFeatureEnabled: boolean;
  isExportsFeatureEnabled: boolean;
  isIbmFeatureEnabled: boolean;
  isNegativeFilteringFeatureEnabled: boolean;
  isOciFeatureEnabled: boolean;
  isUnallocatedCostsFeatureEnabled: boolean;
}>;

export const defaultState: FeatureFlagsState = {
  hasFeatureFlags: false,
  isCostDistributionFeatureEnabled: false,
  isCostTypeFeatureEnabled: false,
  isCurrencyFeatureEnabled: false,
  isDefaultProjectsFeatureEnabled: false,
  isExportsFeatureEnabled: false,
  isIbmFeatureEnabled: false,
  isNegativeFilteringFeatureEnabled: false,
  isOciFeatureEnabled: false,
  isUnallocatedCostsFeatureEnabled: false,
};

export const stateKey = 'featureFlags';

export function featureFlagsReducer(state = defaultState, action: FeatureFlagsAction): FeatureFlagsState {
  switch (action.type) {
    case getType(setFeatureFlags):
      return {
        ...state,
        hasFeatureFlags: true,
        isCostDistributionFeatureEnabled: action.payload.isCostDistributionFeatureEnabled,
        isCostTypeFeatureEnabled: action.payload.isCostTypeFeatureEnabled,
        isCurrencyFeatureEnabled: action.payload.isCurrencyFeatureEnabled,
        isDefaultProjectsFeatureEnabled: action.payload.isDefaultProjectsFeatureEnabled,
        isExportsFeatureEnabled: action.payload.isExportsFeatureEnabled,
        isIbmFeatureEnabled: action.payload.isIbmFeatureEnabled,
        isNegativeFilteringFeatureEnabled: action.payload.isNegativeFilteringFeatureEnabled,
        isOciFeatureEnabled: action.payload.isOciFeatureEnabled,
        isUnallocatedCostsFeatureEnabled: action.payload.isUnallocatedCostsFeatureEnabled,
      };

    default:
      return state;
  }
}
