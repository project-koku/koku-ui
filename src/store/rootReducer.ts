import { notifications } from '@redhat-cloud-services/frontend-components-notifications';
import { combineReducers } from 'redux';
import { awsCostOverviewReducer, awsCostOverviewStateKey } from 'store/breakdown/costOverview/awsCostOverview';
import { azureCostOverviewReducer, azureCostOverviewStateKey } from 'store/breakdown/costOverview/azureCostOverview';
import { gcpCostOverviewReducer, gcpCostOverviewStateKey } from 'store/breakdown/costOverview/gcpCostOverview';
import { ibmCostOverviewReducer, ibmCostOverviewStateKey } from 'store/breakdown/costOverview/ibmCostOverview';
import { ocpCostOverviewReducer, ocpCostOverviewStateKey } from 'store/breakdown/costOverview/ocpCostOverview';
import { awsHistoricalDataReducer, awsHistoricalDataStateKey } from 'store/breakdown/historicalData/awsHistoricalData';
import {
  azureHistoricalDataReducer,
  azureHistoricalDataStateKey,
} from 'store/breakdown/historicalData/azureHistoricalData';
import { gcpHistoricalDataReducer, gcpHistoricalDataStateKey } from 'store/breakdown/historicalData/gcpHistoricalData';
import { ibmHistoricalDataReducer, ibmHistoricalDataStateKey } from 'store/breakdown/historicalData/ibmHistoricalData';
import { ocpHistoricalDataReducer, ocpHistoricalDataStateKey } from 'store/breakdown/historicalData/ocpHistoricalData';
import { costModelsReducer, costModelsStateKey } from 'store/costModels';
import { awsDashboardReducer, awsDashboardStateKey } from 'store/dashboard/awsDashboard';
import { awsOcpDashboardReducer, awsOcpDashboardStateKey } from 'store/dashboard/awsOcpDashboard';
import { azureDashboardReducer, azureDashboardStateKey } from 'store/dashboard/azureDashboard';
import { azureOcpDashboardReducer, azureOcpDashboardStateKey } from 'store/dashboard/azureOcpDashboard';
import { gcpDashboardReducer, gcpDashboardStateKey } from 'store/dashboard/gcpDashboard';
import { gcpOcpDashboardReducer, gcpOcpDashboardStateKey } from 'store/dashboard/gcpOcpDashboard';
import { ibmDashboardReducer, ibmDashboardStateKey } from 'store/dashboard/ibmDashboard';
import { ocpCloudDashboardReducer, ocpCloudDashboardStateKey } from 'store/dashboard/ocpCloudDashboard';
import { ocpDashboardReducer, ocpDashboardStateKey } from 'store/dashboard/ocpDashboard';
import {
  ocpInfrastructureDashboardReducer,
  ocpInfrastructureDashboardStateKey,
} from 'store/dashboard/ocpInfrastructureDashboard';
import {
  ocpSupplementaryDashboardReducer,
  ocpSupplementaryDashboardStateKey,
} from 'store/dashboard/ocpSupplementaryDashboard';
import { ocpUsageDashboardReducer, ocpUsageDashboardStateKey } from 'store/dashboard/ocpUsageDashboard';
import { exportReducer, exportStateKey } from 'store/exports';
import { forecastReducer, forecastStateKey } from 'store/forecasts';
import { orgReducer, orgStateKey } from 'store/orgs';
import { priceListReducer, priceListStateKey } from 'store/priceList';
import { reportReducer, reportStateKey } from 'store/reports';
import { resourceReducer, resourceStateKey } from 'store/resources';
import { sourcesReducer, sourcesStateKey } from 'store/sourceSettings';
import { tagReducer, tagStateKey } from 'store/tags';
import { StateType } from 'typesafe-actions';

import { metricsReducer, metricsStateKey } from './metrics';
import { providersReducer, providersStateKey } from './providers';
import { rbacReducer, rbacStateKey } from './rbac';
import { uiReducer, uiStateKey } from './ui';
import { userAccessReducer, userAccessStateKey } from './userAccess';

export type RootState = StateType<typeof rootReducer>;

export const rootReducer = combineReducers({
  [awsOcpDashboardStateKey]: awsOcpDashboardReducer,
  [awsCostOverviewStateKey]: awsCostOverviewReducer,
  [awsDashboardStateKey]: awsDashboardReducer,
  [awsHistoricalDataStateKey]: awsHistoricalDataReducer,
  [azureOcpDashboardStateKey]: azureOcpDashboardReducer,
  [azureCostOverviewStateKey]: azureCostOverviewReducer,
  [azureDashboardStateKey]: azureDashboardReducer,
  [azureHistoricalDataStateKey]: azureHistoricalDataReducer,
  [costModelsStateKey]: costModelsReducer,
  [exportStateKey]: exportReducer,
  [gcpCostOverviewStateKey]: gcpCostOverviewReducer,
  [gcpDashboardStateKey]: gcpDashboardReducer,
  [gcpOcpDashboardStateKey]: gcpOcpDashboardReducer,
  [gcpHistoricalDataStateKey]: gcpHistoricalDataReducer,
  [ibmCostOverviewStateKey]: ibmCostOverviewReducer,
  [ibmDashboardStateKey]: ibmDashboardReducer,
  [ibmHistoricalDataStateKey]: ibmHistoricalDataReducer,
  [metricsStateKey]: metricsReducer,
  [ocpCostOverviewStateKey]: ocpCostOverviewReducer,
  [ocpDashboardStateKey]: ocpDashboardReducer,
  [ocpCloudDashboardStateKey]: ocpCloudDashboardReducer,
  [ocpHistoricalDataStateKey]: ocpHistoricalDataReducer,
  [ocpInfrastructureDashboardStateKey]: ocpInfrastructureDashboardReducer,
  [ocpSupplementaryDashboardStateKey]: ocpSupplementaryDashboardReducer,
  [ocpUsageDashboardStateKey]: ocpUsageDashboardReducer,
  [orgStateKey]: orgReducer,
  [priceListStateKey]: priceListReducer,
  [providersStateKey]: providersReducer,
  [rbacStateKey]: rbacReducer,
  [reportStateKey]: reportReducer,
  [forecastStateKey]: forecastReducer,
  [resourceStateKey]: resourceReducer,
  [sourcesStateKey]: sourcesReducer,
  [tagStateKey]: tagReducer,
  [uiStateKey]: uiReducer,
  [userAccessStateKey]: userAccessReducer,
  notifications,
});
