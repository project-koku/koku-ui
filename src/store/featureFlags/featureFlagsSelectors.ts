import type { RootState } from 'store/rootReducer';

import { stateKey } from './featureFlagsReducer';

export const selectFeatureFlagsState = (state: RootState) => state[stateKey];

export const selectHasFeatureFlags = (state: RootState) => selectFeatureFlagsState(state).hasFeatureFlags;

export const selectIsCurrencyFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state)?.isCurrencyFeatureEnabled;
export const selectIsCostDistributionFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state)?.isCostDistributionFeatureEnabled;
export const selectIsCostTypeFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isCostTypeFeatureEnabled;
export const selectIsExportsFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isExportsFeatureEnabled;
export const selectIsIbmFeatureEnabled = (state: RootState) => selectFeatureFlagsState(state).isIbmFeatureEnabled;
export const selectIsNegativeFilteringFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isNegativeFilteringFeatureEnabled;
export const selectIsOciFeatureEnabled = (state: RootState) => selectFeatureFlagsState(state).isOciFeatureEnabled;
export const selectIsPlatformCostsFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isPlatformCostsFeatureEnabled;
export const selectIsUnallocatedCostsFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isUnallocatedCostsFeatureEnabled;
