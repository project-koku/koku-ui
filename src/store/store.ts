import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import axios from 'axios';
import { applyMiddleware, compose, createStore, DeepPartial } from 'redux';
import thunk from 'redux-thunk';

import { rootReducer, RootState } from './rootReducer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [thunk, notificationsMiddleware()];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState as any, enhancer);

  axios.interceptors.response.use(null, error => {
    return Promise.reject(error);
  });

  return store;
}
