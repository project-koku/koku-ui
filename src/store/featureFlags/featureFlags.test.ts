import { createMockStoreCreator } from 'store/mockStore';

import { featureFlagsSelectors } from '.';
import * as actions from './featureFlagsActions';
import { featureFlagsReducer, stateKey } from './featureFlagsReducer';
import * as selectors from './featureFlagsSelectors';

const createUIStore = createMockStoreCreator({
  [stateKey]: featureFlagsReducer,
});

test('default state', async () => {
  const store = createUIStore();
  expect(selectors.selectFeatureFlagsState(store.getState())).toMatchSnapshot();
});

test('currency feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isCurrencyFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsCurrencyFeatureEnabled(store.getState())).toBe(true);
});

test('cost distribution feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isCostDistributionFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsCostDistributionFeatureEnabled(store.getState())).toBe(true);
});

test('cost type feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isCostTypeFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsCostTypeFeatureEnabled(store.getState())).toBe(true);
});

test('exports feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isExportsFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsExportsFeatureEnabled(store.getState())).toBe(true);
});

test('IBM feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isIbmFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsIbmFeatureEnabled(store.getState())).toBe(true);
});

test('negative filtering feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isNegativeFilteringFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsNegativeFilteringFeatureEnabled(store.getState())).toBe(true);
});

test('OCI feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isOciFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsOciFeatureEnabled(store.getState())).toBe(true);
});

test('platform costs feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isPlatformCostsFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsPlatformCostsFeatureEnabled(store.getState())).toBe(true);
});

test('unallocated costs feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isUnallocatedCostsFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsUnallocatedCostsFeatureEnabled(store.getState())).toBe(true);
});
