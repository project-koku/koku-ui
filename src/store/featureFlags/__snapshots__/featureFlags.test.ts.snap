// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
{
  "hasFeatureFlags": false,
  "isCostDistributionFeatureEnabled": false,
  "isCostTypeFeatureEnabled": false,
  "isCurrencyFeatureEnabled": false,
  "isDefaultProjectsFeatureEnabled": false,
  "isExportsFeatureEnabled": false,
  "isIbmFeatureEnabled": false,
  "isNegativeFilteringFeatureEnabled": false,
  "isOciFeatureEnabled": false,
  "isUnallocatedCostsFeatureEnabled": false,
}
`;
